keyshare.go

```diff
@@ -1,12 +1,208 @@
+	bts, err := keyshareUserCommitmentsHash(hashInput)
+	if err != nil {
+		return KeyshareCommitmentRequest{}, nil, err
+	}
+	return KeyshareCommitmentRequest{HashedUserCommitments: bts}, hashInput, nil
+}
+
+// KeyshareUserResponseRequest computes the user's second message to the keyshare server in the
+// keyshare protocol, containing its response in the joint computation of the zero- knowledgeproof
+// of the secret key. Also returns the challenge to be used in constructing the proofs.
+func KeyshareUserResponseRequest[T comparable](
+	builders ProofBuilderList,
+	randomizers map[string]*big.Int,
+	hashInput []KeyshareUserChallengeInput[T],
+	context, nonce *big.Int,
+	signature bool,
+) (KeyshareResponseRequest[T], *big.Int, error) {
+	// Extract the user secret from the builders. Since this secret will be the same across all
+	// builders, we can just take it off the first one.
+	// (We extract it manually like this instead of adding a method to the ProofBuilder interface,
+	// because we don't want to expose a method to retrieve the secret in the gabi public API.)
+	var userSecret *big.Int
+	builder := builders[0]
+	switch b := builder.(type) {
+	case *CredentialBuilder:
+		userSecret = b.secret
+	case *DisclosureProofBuilder:
+		userSecret = b.attributes[0]
+	default:
+		return KeyshareResponseRequest[T]{}, nil, errors.New("Unsupported proof builder")
+	}
+
+	challenge, err := builders.ChallengeWithRandomizers(context, nonce, randomizers, signature)
+	if err != nil {
+		return KeyshareResponseRequest[T]{}, nil, err
+	}
+	userResponse := new(big.Int).Add(randomizers["secretkey"], new(big.Int).Mul(challenge, userSecret))
+
+	return KeyshareResponseRequest[T]{
+		Nonce:              nonce,
+		UserResponse:       userResponse,
+		IsSignatureSession: signature,
+		UserChallengeInput: hashInput,
+	}, challenge, nil
+}
+
+func keyshareUserCommitmentsHash[T any](i []KeyshareUserChallengeInput[T]) ([]byte, error) {
+	bts, err := cbor.Marshal(i, cbor.EncOptions{})
+	if err != nil {
+		return nil, err
+	}
+	h := sha256.Sum256(bts)
+	return h[:], nil
+}
+
+// KeyshareResponse generates the keyshare response, using the keyshare secret and the user's
+// input in the keyshare protocol so far.
+func KeyshareResponse[T comparable](
+	secret *big.Int,
+	randomizer *big.Int,
+	req KeyshareCommitmentRequest,
+	res KeyshareResponseRequest[T],
+	keys map[T]*gabikeys.PublicKey,
+) (*ProofP, error) {
+	// Sanity checks
+	for i, k := range res.UserChallengeInput {
+		if k.KeyID != nil && keys[*k.KeyID] == nil {
+			return nil, errors.Errorf("missing public key for element %d of challenge input", i)
+		}
+	}
+	if res.Context == nil {
+		res.Context = bigOne
+	}
+
+	hashContribs := make([]KeyshareUserChallengeInput[T], 0, len(res.UserChallengeInput))
+	challengeContribs := make([]*big.Int, 0, len(res.UserChallengeInput)*2)
+
+	// Assemble the input for the computation of h_W
+	for _, data := range res.UserChallengeInput {
+		hashContribs = append(hashContribs, data)
```

Comment:
Couldn't you just use `res.UserChallengeInput` instead of appending to the new `hashContribs` for the `h_w` input?

Reply:
Yes indeed, thanks for that 😅 I think `hashContribs` is a leftover from an early version of this branch in which it wasn't an exact copy of `res.UserChallengeInput`. But it is now, so I've removed it as you suggested.